File: euler-011.py
#!/usr/bin/env python3

import functools

# loadProblem :: IO [[Int]]
def loadProblem():
  return [[int(n) for n in line.split(' ')] for line in open("euler-011.txt").readlines()]

# getHoriz :: Int -> Int -> [[a]] -> [a]
def getHoriz(row, col, xss):
    [a, b, c, d] = [xss[row][col + i] for i in range(4)]
    return [a, b, c, d]

# getVert :: Int -> Int -> [[a]] -> [a]
def getVert(row, col, xss):
    [a, b, c, d] = [xss[row + i][col] for i in range(4)]
    return [a, b, c, d]

# getSlash :: Int -> Int -> [[a]] -> [a]
def getSlash(row, col, xss):
    [a, b, c, d] = [xss[row + 3 - i][col + i] for i in range(4)]
    return [a, b, c, d]

# getBackslash :: Int -> Int -> [[a]] -> [a]
def getBackslash(row, col, xss):
    [a, b, c, d] = [xss[row + i][col + 3 - i] for i in range(4)]
    return [a, b, c, d]

# getFours :: Show a => [[a]] -> [[a]]
def getFours(xss):
    horizes = [getHoriz(row, col, xss) for row in range(20) for col in range(17)]
    verts = [getVert(row, col, xss) for row in range(17) for col in range(20)]
    slashes = [getSlash(row, col, xss) for row in range(17) for col in range(17)]
    backslashes = [getBackslash(row, col, xss) for row in range(17) for col in range(17)]
    fours = [n for l in [horizes, verts, slashes, backslashes] for n in l]
    return fours

# main :: IO ()
def main():
  nums = loadProblem()
  print(max((functools.reduce(lambda x, y: x*y, fours) for fours in getFours(nums))))

if __name__ == '__main__':
    main()

File: euler-011.hs
#! /usr/bin/env runhaskell

import Debug.Trace
import Control.DeepSeq

loadProblem :: IO [[Int]]
loadProblem = do
  fileStr <- readFile "euler-011.txt"
  let fileLined = lines fileStr
      fileWords = fmap words fileLined
      fileInt   = fmap (fmap read) fileWords
  return fileInt

getItem :: Int -> Int -> [[a]] -> a
getItem row col xss = x
  where
    xs = xss !! row
    x  = xs  !! col

getHoriz :: Int -> Int -> [[a]] -> [a]
getHoriz row col xss = [a, b, c, d]
  where
    [a, b, c, d] = [getItem row (col + i) xss | i <- [0..3]]

getVert :: Int -> Int -> [[a]] -> [a]
getVert row col xss = [a, b, c, d]
  where
    [a, b, c, d] = [getItem (row + i) col xss | i <- [0..3]]

getSlash :: Int -> Int -> [[a]] -> [a]
getSlash row col xss = [a, b, c, d]
  where
    [a, b, c, d] = [getItem (row + 3 - i) (col + i) xss | i <- [0..3]]

getBackslash :: Int -> Int -> [[a]] -> [a]
getBackslash row col xss = [a, b, c, d]
  where
    [a, b, c, d] = [getItem (row + i) (col + 3 - i) xss | i <- [0..3]]

getFours :: Show a => [[a]] -> [[a]]
getFours xss = trace (show . map length $ fours) foursConcat
  where
    horizes      = [getHoriz       row col xss | row <- [0..19], col <- [0..16]]
    verts        = [getVert        row col xss | row <- [0..16], col <- [0..19]]
    slashes      = [getSlash       row col xss | row <- [0..16], col <- [0..16]]
    backslashes  = [getBackslash   row col xss | row <- [0..16], col <- [0..16]]
    fours        = [horizes, verts, slashes, backslashes]
    foursConcat  = concat [horizes, verts, slashes, backslashes]

main :: IO ()
main = do
  nums <- loadProblem
  print $ maximum . map product . getFours $ nums
